sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+    pmp_region_index_t region, uint32_t *field_value) {
+  uint32_t cfg_csr_original;
+  if (!pmp_cfg_csr_rw(region, kPmpCsrAccessTypeRead, &cfg_csr_original)) {
+    return kPmpRegionConfigureError;
+  }
+
+  size_t field_index = (region % PMP_CFG_FIELDS_PER_REG) * PMP_CFG_FIELD_WIDTH;
+  bitfield_field32_t pmp_csr_cfg_field = {
+      .mask = PMP_CFG_FIELD_MASK, .index = field_index,
+  };
+
+  *field_value = bitfield_field32_read(cfg_csr_original, pmp_csr_cfg_field);
+
+  return kPmpRegionConfigureOk;
+}
+
+// TODO - description.
+static pmp_region_configure_result_t pmp_csr_cfg_field_write(
+    pmp_region_index_t region, uint32_t field_value) {
+  uint32_t cfg_csr_current;
+  if (!pmp_cfg_csr_rw(region, kPmpCsrAccessTypeRead, &cfg_csr_current)) {
+    return kPmpRegionConfigureError;
+  }
+
+  // Determine the pmpcfg field index based on the `region`.
+  size_t field_index = (region % PMP_CFG_FIELDS_PER_REG) * PMP_CFG_FIELD_WIDTH;
+  bitfield_field32_t pmp_csr_cfg_field = {
+      .mask = PMP_CFG_FIELD_MASK, .index = field_index,
+  };
+
+  uint32_t cfg_csr_new =
+      bitfield_field32_write(cfg_csr_current, pmp_csr_cfg_field, field_value);
+
+  if (!pmp_cfg_csr_rw(region, kPmpCsrAccessTypeWrite, &cfg_csr_new)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (!pmp_cfg_csr_rw(region, kPmpCsrAccessTypeRead, &cfg_csr_current)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (cfg_csr_current != cfg_csr_new) {
+    return kPmpRegionConfigureWarlError;
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+/**
+ * Writes `address` to a pmpaddr CSRs.
+ *
+ * The corresponding pmpaddrN index N is determined by `region`.
+ *
+ * PMP address must be at least 4bytes aligned, and pmpaddr holds only bits
+ * 33:2. This means that before writing an address to a pmpaddr CSR, it must be
+ * shifted 2 bits to the right.
+ *
+ * Please see:
+ * "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+ * "3.6.1 Physical Memory Protection CSRs",
+ * "Figure 3.26: PMP address register format, RV32".
+ *
+ * @param region PMP region to configure and set address for.
+ * @param address PMP address to be set.
+ * @return `pmp_region_configure_result_t`.
+ */
+pmp_region_configure_result_t pmp_csr_address_write(
+    pmp_region_index_t region, pmp_region_address_t address) {
+  pmp_region_address_t address_shifted = address >> PMP_ADDRESS_SHIFT;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeWrite, &address_shifted)) {
+    return kPmpRegionConfigureError;
+  }
+
+  pmp_region_address_t addr_csr_after_write;
+  if (!pmp_addr_csr_rw(region, kPmpCsrAccessTypeRead, &addr_csr_after_write)) {
+    return kPmpRegionConfigureError;
+  }
+
+  if (address_shifted != addr_csr_after_write) {
+    return kPmpRegionConfigureWarlError;
+  }
+
+  return kPmpRegionConfigureOk;
+}
+
+// TODO - short description.
+static bool pmp_cfg_permissions_set(pmp_region_permissions_t perm,
+                                    uint32_t *bitfield) {
+  switch (perm) {
```

Comment:
This makes me feel that `perm` should be a bitflag value, rather than an enum. It would definitely simply the code (and, for callers, allow the nicer `R | X`-type API).

Reply:
True, I guess I can just check for an illegal combination inside this function.

Reply:
Leaving this decision for the follow-up.